wgpu-core/src/weak_vec.rs

```diff
@@ -0,0 +1,66 @@
+use std::sync::Weak;
+
+/// A container that holds Weak references of T.
+///
+/// On `push` it scans its contents for weak references with no strong references still alive and drops them.
+#[derive(Debug)]
+pub(crate) struct WeakVec<T> {
+    inner: Vec<Option<Weak<T>>>,
+    empty_slots: Vec<usize>,
+}
+
+impl<T> Default for WeakVec<T> {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+            empty_slots: Default::default(),
+        }
+    }
+}
+
+impl<T> WeakVec<T> {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: Vec::new(),
+            empty_slots: Vec::new(),
+        }
+    }
+
+    pub(crate) fn push(&mut self, value: Weak<T>) {
+        // Remove stale weak references
+        for (i, value) in self.inner.iter_mut().enumerate() {
+            if let Some(w) = value {
+                if w.strong_count() == 0 {
+                    *value = None;
+                    self.empty_slots.push(i);
+                }
+            }
+        }
+        if let Some(i) = self.empty_slots.pop() {
+            self.inner[i] = Some(value);
+        } else {
+            self.inner.push(Some(value));
+        }
```

Comment:
Right, good point.

Reply:
Hmm, I was in process of addressing this with:

```rust
pub(crate) fn push(&mut self, value: Weak<T>) {
    let mut to_insert = Some(value);
    for slot in &mut self.inner {
        if let Some(w) = slot {
            if w.strong_count() == 0 {
                *slot = to_insert.take();
            }
        } else {
            *slot = to_insert.take();
        }
    }
    if let Some(to_insert) = to_insert {
        self.inner.push(Some(to_insert));
    }
}
```

Comparing the complied code it looks like this function has less jumps and moves.
Probably because the approach in https://github.com/gfx-rs/wgpu/commit/1f94aad66aa777b0adb12c2e7756012cd3d5f295 is not as straight forward to the compiler?

https://rust.godbolt.org/z/zKrE4xPKK
https://rust.godbolt.org/z/W7nW68WPW

https://www.diffchecker.com/XWsoSpeZ/

Reply:
I will revise the patch to use the code above.